openwisp_network_topology/consumers.py

```diff
@@ -2,31 +2,35 @@
 
 from asgiref.sync import async_to_sync
 from channels.generic.websocket import WebsocketConsumer
+from swapper import load_model
+
+Topology = load_model('topology', 'Topology')
 
 
 class TopologyConsumer(WebsocketConsumer):
     channel_layer_group = 'topology'
 
-    def _is_user_authenticated(self):
-        try:
-            assert self.scope['user'].is_authenticated is True
-        except (KeyError, AssertionError):
-            self.close()
-            return False
-        else:
+    def _is_user_authorized_to_view_topology(self, user, topology_pk):
+        if user.is_authenticated and user.is_superuser:
             return True
+        topology = Topology.objects.get(pk=topology_pk)
+        return (
+            user.is_authenticated
+            and user.has_perm(f'{Topology._meta.app_label}.view_topology')
+            and user.is_manager(topology.organization.pk)
```

Comment:
Changing `topology.organization.pk` to `topology.organization_id` won't hit the DB in case the `organization` object is not fetched yet, we do this often in our codebase.

Let's also execute the org membership check first.